### PR TITLE
add code

### PR DESCRIPTION
**Pull request by zekun-li**

### CODE REVIEW

**Comment by SasanTV** [comment]
Could you please move your codes under a directory? You can name the directory after yourself, or a cool name for your implementation :)
The idea is that we may have several independent implementations for each module. We'd like to host them under one repositories but different directories.

File: step1_pred_pairwise_novgg_withcap.py [hunk 1]
```diff
@@ -0,0 +1,470 @@
...
+# configure GPU usage
+from keras.backend.tensorflow_backend import set_session
+config = tf.ConfigProto()
+config.gpu_options.per_process_gpu_memory_fraction = 0.4
+set_session(tf.Session(config=config))
+
+
+def parse_cmdline_args():
+    parser = argparse.ArgumentParser(description='Parser for deep metric learning prediction code')
+    parser.add_argument('--preprocess_dir', type = str, default = '/data/zekunl/Geolocalizer/google_grouping/')
```
Comment (SasanTV) [inline step1_pred_pairwise_novgg_withcap.py:36]: Please add "help='some explanation'" to all arguments.

File: step1_pred_pairwise_novgg_withcap.py [hunk 2]
```diff
@@ -0,0 +1,470 @@
...
+                        default = '/data/zekunl/text_linking/style_learner/capitalization_real.hdf5')
+    
+    parser.add_argument('--ALPHA', type = int, default = 1)
+    parser.add_argument('--max_batch_size', type = int, default = 256)
+    return (parser.parse_args())
+    
+
+# code for removing special caracters, puctuations
+# https://mlwhiz.com/blog/2019/01/17/deeplearning_nlp_preprocess/
+puncts = [',', '.', '"', ':', ')', '(', '-', '!', '?', '|', ';', "'", '$', '&', '/', '[', ']', '>', '%', '=', '#', '*', '+', '\\', '•',  '~', '@', '£',
```
Comment (SasanTV) [inline step1_pred_pairwise_novgg_withcap.py:54]: I prefer regex, but your call.

File: step1_pred_pairwise_novgg_withcap.py [hunk 3]
```diff
@@ -0,0 +1,470 @@
...
+
+# code for removing special caracters, puctuations
+# https://mlwhiz.com/blog/2019/01/17/deeplearning_nlp_preprocess/
+puncts = [',', '.', '"', ':', ')', '(', '-', '!', '?', '|', ';', "'", '$', '&', '/', '[', ']', '>', '%', '=', '#', '*', '+', '\\', '•',  '~', '@', '£', 
+ '·', '_', '{', '}', '©', '^', '®', '`',  '<', '→', '°', '€', '™', '›',  '♥', '←', '×', '§', '″', '′', 'Â', '█', '½', 'à', '…', 
+ '“', '★', '”', '–', '●', 'â', '►', '−', '¢', '²', '¬', '░', '¶', '↑', '±', '¿', '▾', '═', '¦', '║', '―', '¥', '▓', '—', '‹', '─', 
+ '▒', '：', '¼', '⊕', '▼', '▪', '†', '■', '’', '▀', '¨', '▄', '♫', '☆', 'é', '¯', '♦', '¤', '▲', 'è', '¸', '¾', 'Ã', '⋅', '‘', '∞', 
+ '∙', '）', '↓', '、', '│', '（', '»', '，', '♪', '╩', '╚', '³', '・', '╦', '╣', '╔', '╗', '▬', '❤', 'ï', 'Ø', '¹', '≤', '‡', '√', ]
+
+def clean_text(x):
```
Comment (SasanTV) [inline step1_pred_pairwise_novgg_withcap.py:60]: clean_symbols?

File: step1_pred_pairwise_novgg_withcap.py (hunk 2 above)
Comment (SasanTV) [inline step1_pred_pairwise_novgg_withcap.py:54]: symbols?

File: step1_pred_pairwise_novgg_withcap.py [hunk 4]
```diff
@@ -0,0 +1,470 @@
...
+                      padding='same',
+                      name='block5_conv1')(x)
+    
+    x = layers.MaxPooling2D((2, 2), strides=(2, 2), name='block5_pool')(x)
+
+
+    # Classification block
+    x = layers.Flatten(name='flatten')(x)
+    x = layers.Dense(1024, activation='relu', name='fc1')(x)
+    x = layers.Dense(500, activation='softmax', name='predictions')(x)
```
Comment (SasanTV) [inline step1_pred_pairwise_novgg_withcap.py:142]: Why 500?

File: step1_pred_pairwise_novgg_withcap.py [hunk 5]
```diff
@@ -0,0 +1,470 @@
...
+
+def trip_feat_loss_shape(shapes):
+    shape1, shape2, shape3 = shapes
+    return (shape1[0], 1)
+
+    
+def my_visual_model():
+    img_input = Input(shape=(224,224,3))
+    
+    x = layers.Conv2D(32, (3, 3),
```
Comment (SasanTV) [inline step1_pred_pairwise_novgg_withcap.py:99]: Is there a reason you did not choose one of the available architectures?

File: step1_pred_pairwise_novgg_withcap.py [hunk 6]
```diff
@@ -0,0 +1,470 @@
...
+
+def feature_model():
+    input_text = Input(shape = (50,))
+    input_loc = Input(shape = (4,))
+    
+    t = input_text
+    
+    l = input_loc
+    
+    x = keras.layers.Concatenate(axis=-1)([ t, l])
```
Comment (SasanTV) [inline step1_pred_pairwise_novgg_withcap.py:156]: mix_representation = keras.layers.Concatenate(axis=-1)([ input_text, input_loc]) ?

File: step1_pred_pairwise_novgg_withcap.py [hunk 7]
```diff
@@ -0,0 +1,470 @@
...
+    
+    model = keras.models.Model([img_input], [x], name='my_visual_model')
+    
+    return model
+
+def feature_model():
+    input_text = Input(shape = (50,))
+    input_loc = Input(shape = (4,))
+    
+    t = input_text
```
Comment (SasanTV) [inline step1_pred_pairwise_novgg_withcap.py:152]: why do you need t and l?

File: step1_pred_pairwise_novgg_withcap.py [hunk 8]
```diff
@@ -0,0 +1,470 @@
...
+    model = keras.models.Model([input_text, input_loc], mix_representation)
+    print( model.summary())
+    
+    return model
+
+    
+def clf_model():
+    
+    model = Sequential()
+    model.add(Dense(128, input_shape=( ( 50 + 2 + 2 +2 ) * 2,), activation = 'relu'))
```
Comment (SasanTV) [inline step1_pred_pairwise_novgg_withcap.py:169]: Please add a comment to explain the input shape.

File: step1_pred_pairwise_novgg_withcap.py [hunk 9]
```diff
@@ -0,0 +1,470 @@
...
+
+    return loss
+
+def trip_feat_loss_shape(shapes):
+    shape1, shape2, shape3 = shapes
+    return (shape1[0], 1)
+
+    
+def my_visual_model():
+    img_input = Input(shape=(224,224,3))
```
Comment (SasanTV) [inline step1_pred_pairwise_novgg_withcap.py:97]: Please add a comment to explain the input shape.

File: step1_pred_pairwise_novgg_withcap.py [hunk 10]
```diff
@@ -0,0 +1,470 @@
...
+    
+    cap_model = keras.models.load_model(cap_weights)
+    for layer in cap_model.layers:
+        layer.trainable = False
+
+
+    cap_feat1 = cap_model(input_visual1)
+    cap_feat2 = cap_model(input_visual2)
+
+    input_text1 = Input(shape = (50,))
```
Comment (SasanTV) [inline step1_pred_pairwise_novgg_withcap.py:192]: Please add comment to explain the input shape.

File: step1_pred_pairwise_novgg_withcap.py [hunk 11]
```diff
@@ -0,0 +1,470 @@
...
+    input_text2 = Input(shape = (50,))
+    
+    input_loc1 = Input(shape = (4,))
```
Comment (SasanTV) [inline step1_pred_pairwise_novgg_withcap.py:195]: ditto

File: step1_pred_pairwise_novgg_withcap.py [hunk 12]
```diff
@@ -0,0 +1,470 @@
...
+            anchor_angle = angle_array[anchor_idx]
+            anchor_fontsize = font_size_array[anchor_idx]
+
+
+
+            OV = []
+            OT = []
+            OL = []
+            OA = []
+            OF = []
```
Comment (SasanTV) [inline step1_pred_pairwise_novgg_withcap.py:392]: Please add comments, or use full names (full names are preferred).

File: step1_pred_pairwise_novgg_withcap.py [hunk 13]
```diff
@@ -0,0 +1,470 @@
...
+
+            Y_pred = []
+            pred_values = []
+
+            for other_idx in range(total_num_regions):
```
Comment (SasanTV) [inline step1_pred_pairwise_novgg_withcap.py:397]: what does "other" refer to?

File: step1_pred_pairwise_novgg_withcap.py [hunk 14]
```diff
@@ -0,0 +1,470 @@
...
+                if (other_idx + 1 )% max_batch_size == 0 or (other_idx) == (total_num_regions -1 ): # Time to evaluate!
+
+                    OV = np.array(OV)
+                    OT = np.array(OT)
+                    OL = np.array(OL)
+                    OA = np.array(OA)
+                    OF = np.array(OF)
+                    OL = np.concatenate([OL, OA, OF], axis = 1)
+
+                    XV = np.repeat(anchor_img[np.newaxis, :], OV.shape[0], axis = 0 )
```
Comment (SasanTV) [inline step1_pred_pairwise_novgg_withcap.py:426]: ditto

File: step1_pred_pairwise_novgg_withoutcap.py [hunk 15]
```diff
@@ -0,0 +1,458 @@
+# coding: utf-8
```
Comment (SasanTV) [inline step1_pred_pairwise_novgg_withoutcap.py:1]: Seems that most of the functions here and in _withcap.py are the same. Please refactor the identical functions into a common library (file).
Ideally, cap and no_cap should be a flag given to the same file, but if that's too much work, at least lets refactor the common functions.

**Review by SasanTV — CHANGES_REQUESTED** [review]

File: step1_pred_pairwise_novgg_withcap.py (hunk 1 above)
Comment (zekun-li) [reply step1_pred_pairwise_novgg_withcap.py:36]: Added

File: step1_pred_pairwise_novgg_withcap.py (hunk 2 above)
Comment (zekun-li) [reply step1_pred_pairwise_novgg_withcap.py:54]: modified

File: step1_pred_pairwise_novgg_withcap.py (hunk 4 above)
Comment (zekun-li) [reply step1_pred_pairwise_novgg_withcap.py:142]: This module was not used. Deleted.

File: step1_pred_pairwise_novgg_withcap.py (hunk 6 above)
Comment (zekun-li) [reply step1_pred_pairwise_novgg_withcap.py:156]: Changed.

File: step1_pred_pairwise_novgg_withcap.py (hunk 7 above)
Comment (zekun-li) [reply step1_pred_pairwise_novgg_withcap.py:152]: Changed.

File: step1_pred_pairwise_novgg_withcap.py (hunk 8 above)
Comment (zekun-li) [reply step1_pred_pairwise_novgg_withcap.py:169]: Added Comments

File: step1_pred_pairwise_novgg_withcap.py (hunk 9 above)
Comment (zekun-li) [reply step1_pred_pairwise_novgg_withcap.py:97]: Added comment

File: step1_pred_pairwise_novgg_withcap.py (hunk 10 above)
Comment (zekun-li) [reply step1_pred_pairwise_novgg_withcap.py:192]: Added comment

File: step1_pred_pairwise_novgg_withcap.py (hunk 11 above)
Comment (zekun-li) [reply step1_pred_pairwise_novgg_withcap.py:195]: Added comment

File: step1_pred_pairwise_novgg_withcap.py (hunk 12 above)
Comment (zekun-li) [reply step1_pred_pairwise_novgg_withcap.py:392]: Added comment

File: step1_pred_pairwise_novgg_withcap.py (hunk 13 above)
Comment (zekun-li) [reply step1_pred_pairwise_novgg_withcap.py:397]: Explained with comment

File: step1_pred_pairwise_novgg_withcap.py (hunk 14 above)
Comment (zekun-li) [reply step1_pred_pairwise_novgg_withcap.py:426]: Explained in the comment

File: step1_pred_pairwise_novgg_withoutcap.py (hunk 15 above)
Comment (zekun-li) [reply step1_pred_pairwise_novgg_withoutcap.py:1]: I've merged two files

**Review by SasanTV — APPROVED** [review]
